js/core.js

```diff
@@ -1162,6 +1162,36 @@ async function parsePhraseBook(s, loadSketch) {
             phraseBook[phrase.key].parameters = phrase.parameters;
         }
     }
+
+    let references = [];
+    function traverse(x){
+        if(x.type === "Concat"){
+            for(key in x){
+                if(x[key].type){
+                    traverse(x[key]);
+                }
+            }
+        }
+        else if(x.type === "Ref"){
+            if(references.indexOf(x.node.key) === -1){
+                references[references.length] = x.node.key;
+            }
+        }
+    }
+    phrases.forEach(function(item){
+        if(item.key === "root"){
+            item.values.forEach(function(i){
+                let p = parsePhrase(i);
```

Comment:
Why are you calling parsePhrase? The source has already been parsed at this point.